ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TxStats.java

```diff
@@ -184,9 +204,37 @@ void incrementAbortedTransactions()
         numberOfAbortedTransactions.incrementAndGet();
 	}
 
+	/**
+	 * @Deprecated as of 5.0.5.Final use {@link #incrementCommittedTransactions(long)}} instead
+	 */
+	@Deprecated
 	void incrementCommittedTransactions()
 	{
-        numberOfCommittedTransactions.incrementAndGet();
+        incrementCommittedTransactions(0L);
+	}
+
+	/**
+	 * @param duration the time in nanoseconds it took for the 2PC phase to complete. The averaged commit
+	 *                   time is available by calling {@link #getAverageCommitTime()}
+	 */
+	void incrementCommittedTransactions(long duration) {
+		TxCommitStatistic prev, next;
+
+		if (duration == 0)
+			duration = 1; // a duration of zero is disallowed - round up to minimum unit
+
+		do {
+			prev = commitStatistic.get();
+
+			long prevCount = prev.getNumberOfCommittedTransactions();
```

Comment:
The average commit time depends upon how many previous commits there were. So when a new transaction is committed the new average will be:

newAverage = (duration  + prevCount * prevAvg) / (prevCount + 1)

where duration is how long it took to commit the new transaction.

Therefore I need to (atomically) get the current number of committed transactions (prevCount) and the current average (prevAvg) in order to calculate the newAverage.

To atomically read (or write) a pair of values we need an AtomicReference (which is quicker that using synchronized blocks).